lib/web/desktop_playback.go

```diff
@@ -17,36 +17,69 @@ limitations under the License.
 package web
 
 import (
+	"context"
 	"net/http"
 
 	"github.com/gravitational/trace"
 	"github.com/julienschmidt/httprouter"
 	"golang.org/x/net/websocket"
 
+	"github.com/gravitational/teleport/lib/player"
 	"github.com/gravitational/teleport/lib/reversetunnelclient"
+	"github.com/gravitational/teleport/lib/session"
 	"github.com/gravitational/teleport/lib/web/desktop"
 )
 
 func (h *Handler) desktopPlaybackHandle(
 	w http.ResponseWriter,
 	r *http.Request,
 	p httprouter.Params,
-	ctx *SessionContext,
+	sctx *SessionContext,
 	site reversetunnelclient.RemoteSite,
 ) (interface{}, error) {
 	sID := p.ByName("sid")
 	if sID == "" {
-		return nil, trace.BadParameter("missing sid in request URL")
+		return nil, trace.BadParameter("missing session ID in request URL")
 	}
 
-	clt, err := ctx.GetUserClient(r.Context(), site)
+	clt, err := sctx.GetUserClient(r.Context(), site)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	websocket.Handler(func(ws *websocket.Conn) {
-		defer h.log.Debug("desktopPlaybackHandle websocket handler returned")
-		desktop.NewPlayer(sID, ws, clt, h.log).Play(r.Context())
+		ws.PayloadType = websocket.BinaryFrame
+
+		player, err := player.New(&player.Config{
+			Clock:     h.clock,
+			Log:       h.log,
+			SessionID: session.ID(sID),
+			Streamer:  clt,
+		})
+		if err != nil {
+			h.log.Errorf("couldn't create player for session %v: %v", sID, err)
+			ws.Write([]byte(`{"message": "error", "errorText": "Internal server error"}`))
+			return
+		}
+
+		defer player.Close()
+
+		ctx, cancel := context.WithCancel(r.Context())
+		defer cancel()
```

Comment:
The context here is used to cancel the goroutines that are launched here.

The player itself doesn't impose how you use it and when/where you launch goroutines. The idea is to keep the code related to terminating goroutines as close as possible to the code that launches them.

`player.Close()` ensures goroutines internal to the player properly terminate.

I'm not sure why we would expose another channel that gets closed when the player is done. `player.C()` already does exactly that.

Reply:
> The context here is used to cancel the goroutines that are launched here.

I believe the context here will rarely cancel the `ReceivePlaybackActions` goroutine: chances are it will be be hanging [here](https://github.com/gravitational/teleport/blob/124217fac0d534b6f3787da0048a4d4b0a478011/lib/web/desktop/playback.go#L78) waiting to receive input, and thus will return either because the frontend closed the websocket, or because [this `<-ctx.Done()`](https://github.com/gravitational/teleport/blob/124217fac0d534b6f3787da0048a4d4b0a478011/lib/web/desktop_playback.go#L81) receives, thereby returning from the parent `Handler` which also closes the websocket. There's only a very minuscule chance that the stars ever align such that [this](https://github.com/gravitational/teleport/blob/124217fac0d534b6f3787da0048a4d4b0a478011/lib/web/desktop/playback.go#L72-L73) is ever selected.

Hence my suggestion is that we don't even bother with passing in the context there, we just use the closing of the websocket to signal that goroutine to return.

> I'm not sure why we would expose another channel that gets closed when the player is done. player.C() already does exactly that.

The idea there was to use `<-player.Done()` to take the place of what `<-ctx.Done()` is doing currently, which is to say that `player` playback was cancelled externally (as opposed to reaching the end of the recording or failing due to an error, which `player.C()` signifies). Granted that adds additional semantics to `player.done` that aren't what you intended and is a finicky requirement on how to use `player`.

How about just removing the `ctx` from `ReceivePlaybackActions`?

```patch
diff --git a/lib/web/desktop/playback.go b/lib/web/desktop/playback.go
index 42d126d6f8..f2932eec7d 100644
--- a/lib/web/desktop/playback.go
+++ b/lib/web/desktop/playback.go
@@ -58,7 +58,6 @@ type actionMessage struct {
 // ReceivePlaybackActions handles logic for receiving playbackAction messages
 // over the websocket and updating the player state accordingly.
 func ReceivePlaybackActions(
-	ctx context.Context,
 	log logrus.FieldLogger,
 	ws *websocket.Conn,
 	player *player.Player) {
@@ -68,12 +67,6 @@ func ReceivePlaybackActions(
 	playing := true
 
 	for {
-		select {
-		case <-ctx.Done():
-			return
-		default:
-		}
-
 		var action actionMessage
 		if err := websocket.JSON.Receive(ws, &action); err != nil {
 			// Connection close errors are expected if the user closes the tab.
@@ -110,12 +103,19 @@ func PlayRecording(
 	log logrus.FieldLogger,
 	ws *websocket.Conn,
 	player *player.Player) {
+	// Start playing the recording
+	player.Play()
+
+	// Send recording events to the client until the player is done
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case evt, ok := <-player.C():
+			// The player's emit channel was closed, this means either the
+			// end of the recording was reached or an internal player error occurred.
 			if !ok {
+				// If the player has an error, send it to the client
 				if playerErr := player.Err(); playerErr != nil {
 					// Attempt to JSONify the error (escaping any quotes)
 					msg, err := json.Marshal(playerErr.Error())
@@ -129,6 +129,8 @@ func PlayRecording(
 					}
 					return
 				}
+
+				// If the end of the recording was reached, send an end message.
 				if _, err := ws.Write([]byte(`{"message":"end"}`)); err != nil {
 					log.Errorf("failed to write end message: %v", err)
 				}
diff --git a/lib/web/desktop_playback.go b/lib/web/desktop_playback.go
index 8b7dd6be28..6db7898927 100644
--- a/lib/web/desktop_playback.go
+++ b/lib/web/desktop_playback.go
@@ -62,23 +62,27 @@ func (h *Handler) desktopPlaybackHandle(
 			return
 		}
 
+		// Ensure that the player gets cleaned up
 		defer player.Close()
 
 		ctx, cancel := context.WithCancel(r.Context())
-		defer cancel()
 
 		go func() {
+			// Cancelling the ctx ensures
+			// that desktop.PlayRecording() returns
 			defer cancel()
-			desktop.ReceivePlaybackActions(ctx, h.log, ws, player)
+			desktop.ReceivePlaybackActions(h.log, ws, player)
 		}()
 
 		go func() {
-			defer cancel()
-			player.Play()
+			// Calling Close() on the websocket ensures
+			// that desktop.ReceivePlaybackActions() returns
+			defer ws.Close()
 			desktop.PlayRecording(ctx, h.log, ws, player)
 		}()
 
-		<-ctx.Done()
+		// Wait for the player to finish
+		<-player.Done()
 	}).ServeHTTP(w, r)
 
 	return nil, nil
```

Reply:
I can play with that. I agree the context is unlikely to actually cause cancelation here.